src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -338,4 +295,169 @@ std::optional<edm4eic::MutableCluster> CalorimeterClusterRecoCoG::reconstruct(co
+        break;
+      }
+    }  // end sim hit loop
+
+    // if no matching cell ID found, continue
+    // otherwise increment sum and add to vector
+    if (iSimMatch == mchits->size()) {
+      debug("No matching SimHit for hit {}", clhit.getCellID());
+      continue;
+    } else {
+      vecAssocSimHits.push_back(iSimMatch);
+      eSimHitSum += (*mchits)[iSimMatch].getEnergy();
+    }
+#endif
+    trace("{} associated sim hits found for reco hit (cell ID = {})", vecAssocSimHits.size(), clhit.getCellID());
+
+    // ------------------------------------------------------------------------
+    // 2. identify primaries contributing to sim hit and sum their contributions
+    // ------------------------------------------------------------------------
+    for (const int iAssocSimHit : vecAssocSimHits) {
+#if EDM4EIC_VERSION_MAJOR >= 7
+      for (std::size_t iContrib = 0; const auto& contrib : (*mchitassociations)[iAssocSimHit].getSimHit().getContributions()) {
+#else
+      for (std::size_t iContrib = 0; const auto& contrib : (*mchits)[iAssocSimHit].getContributions()) {
+#endif
+        // grab primary responsible for contribution
+        edm4hep::MCParticle primary;
+        get_primary(contrib, primary);
+
+        // increment sums accordingly
+        const int idPrim = primary.getObjectID().index;
+        if (mapMCParToContrib.find(idPrim) == mapMCParToContrib.end()) {
+          mapMCParToSimIndices.insert(
+            {idPrim, std::make_pair(iAssocSimHit, iContrib)}
+          );
+          mapMCParToContrib.insert(
+            {idPrim, contrib.getEnergy()}
+          );
+        } else {
+          mapMCParToContrib[idPrim] += contrib.getEnergy();
+        }
+        trace("Identified primary: id = {}, pid = {}, total energy = {}, contributed = {}",
+          primary.getObjectID().index,
+          primary.getPDG(),
+          primary.getEnergy(),
+          mapMCParToContrib[idPrim]
+        );
+
+      }  // end contribution loop
+    }  // end associated sim hit loop
+  }  // end hit loop
+  debug("Found {} primaries contributing a total of {} GeV", mapMCParToContrib.size(), eSimHitSum);
+
+  // --------------------------------------------------------------------------
+  // 3. create association for each contributing primary
+  // --------------------------------------------------------------------------
+  for (const auto& parAndSimIndices : mapMCParToSimIndices) {
+
+    // grab indices, calculate weight
+    const uint32_t iAssocSimHit = parAndSimIndices.second.first;
+    const uint32_t iContrib = parAndSimIndices.second.second;
+    const double weight = mapMCParToContrib[parAndSimIndices.first] / eSimHitSum;
+
+    // get relevant MCParticle
+    edm4hep::MCParticle primary;
+#if EDM4EIC_VERSION_MAJOR >= 7
+    get_primary((*mchitassociations)[iAssocSimHit].getSimHit().getContributions(iContrib), primary);
+#else
+    get_primary((*mchits)[iAssocSimHit].getContributions(iContrib), primary);
+#endif
```

Comment:
Why do you need to do another lookup, if you have `parAndSimIndices.first`?

Reply:
This was mostly a consequence of not providing `MCParticles` as an input to the algorithm (which -- TBH -- I didn't think to do just because the original code didn't do so). Since I couldn't do something like `(*mcpars)[ parAndSimIndices.first ]`, I instead just had it look the primary up again based on its associated sim hit and contribution indices.

One way this could be simplified might be to key these maps on the proxy objects themselves, which would eliminate the need for the lookup. Or I could provide `MCParticles` as an argument and look it up via the `ObjectID().index`.

Reply:
Technically, MCParticles should be provided, but that's not the point. If you just saved `edm4eic::MCParticles` to the key, that would allow you to access it here.

Reply:
Gotcha! Makes sense!